### ISSUE TITLE
[crypto-autotest/autotest-nightly] [TEST compress_qat] vbdev_compress.c: 440:_reduce_rw_blocks_cb: *ERROR*: status -22 on operation from reduce API

### ISSUE THREAD

**Issue by kgodzwox** [spdk/spdk#2499]
# Bug report
Issue appeared during test execution

vbdev_compress.c: 440:_reduce_rw_blocks_cb: *ERROR*: status -22 on operation from reduce API

**TEST compress_qat**
autotest-nightly/crypto-autotest

# Link to the failed CI build
https://ci.spdk.io/results/autotest-nightly/builds/667/archive/crypto-autotest/build.log
https://ci.spdk.io/results/autotest-nightly/builds/668/archive/crypto-autotest/build.log

# Execution failed at
```
00:11:04.018  Suite: bdevio tests on: COMP_lvs0/lv0
00:11:04.018    Test: blockdev write read block ...passed
00:11:04.018    Test: blockdev write zeroes read block ...passed
00:11:04.277    Test: blockdev write zeroes read no split ...passed
00:11:04.277    Test: blockdev write zeroes read split ...passed
00:11:04.277    Test: blockdev write zeroes read split partial ...passed
00:11:04.278    Test: blockdev reset ...[2022-05-05 09:11:50.001286] vbdev_compress.c: 839:_comp_bdev_io_submit: *ERROR*: Unknown I/O type 5
00:11:04.278  passed
00:11:04.278    Test: blockdev write read 8 blocks ...passed
00:11:04.278    Test: blockdev write read size > 128k ...passed
00:11:04.278    Test: blockdev write read invalid size ...passed
00:11:04.278    Test: blockdev write read offset + nbytes == size of blockdev ...passed
00:11:04.278    Test: blockdev write read offset + nbytes > size of blockdev ...passed
00:11:04.278    Test: blockdev write read max offset ...passed
00:11:04.278    Test: blockdev write read 2 blocks on overlapped address offset ...passed
00:11:04.278    Test: blockdev writev readv 8 blocks ...passed
00:11:04.278    Test: blockdev writev readv 30 x 1block ...[2022-05-05 09:11:50.018830] vbdev_compress.c: 440:_reduce_rw_blocks_cb: *ERROR*: status -22 on operation from reduce API
00:11:04.278  [2022-05-05 09:11:50.018911] vbdev_compress.c: 440:_reduce_rw_blocks_cb: *ERROR*: status -22 on operation from reduce API
00:11:04.278  FAILED
00:11:04.278      1. bdevio.c:464  - CU_ASSERT_EQUAL(g_completion_success,true)
00:11:04.278      2. bdevio.c:471  - CU_ASSERT_EQUAL(g_completion_success,true)
00:11:04.278    Test: blockdev writev readv block ...passed
00:11:04.278    Test: blockdev writev readv size > 128k ...passed
00:11:04.278    Test: blockdev writev readv size > 128k in two iovs ...passed
00:11:04.278    Test: blockdev comparev and writev ...passed
00:11:04.278    Test: blockdev nvme passthru rw ...passed
00:11:04.278    Test: blockdev nvme passthru vendor specific ...passed
00:11:04.278    Test: blockdev nvme admin passthru ...passed
00:11:04.278  
00:11:04.278  Run Summary:    Type  Total    Ran Passed Failed Inactive
00:11:04.278                suites      1      1    n/a      0        0
00:11:04.278                 tests     22     22     21      1        0
00:11:04.278               asserts     89     89     87      2      n/a
00:11:04.278  
00:11:04.278  Elapsed time =    0.539 seconds
00:11:04.278  request:
00:11:04.278  {
00:11:04.278    "method": "perform_tests",
00:11:04.278    "req_id": 1
00:11:04.278  }
00:11:04.278  Got JSON-RPC error response
00:11:04.278  response:
00:11:04.278  {
00:11:04.278    "code": -32603,
00:11:04.278    "message": "2 test cases failed"
00:11:04.278  }
```

```
00:11:04.278  ========== Backtrace start: ==========
00:11:04.278  
00:11:04.278  in ./test/compress/compress.sh:50 -> run_bdevio([])
00:11:04.278       ...
00:11:04.278     45  		$rootdir/test/bdev/bdevio/bdevio -w &
00:11:04.278     46  		bdevio_pid=$!
00:11:04.278     47  		trap 'killprocess $bdevio_pid; error_cleanup; exit 1' SIGINT SIGTERM EXIT
00:11:04.278     48  		waitforlisten $bdevio_pid
00:11:04.278     49  		create_vols
00:11:04.278  => 50  		$rootdir/test/bdev/bdevio/tests.py perform_tests
00:11:04.278     51  		destroy_vols
00:11:04.278     52  		trap - SIGINT SIGTERM EXIT
00:11:04.278     53  		killprocess $bdevio_pid
00:11:04.278     54  	}
00:11:04.278     55  	
00:11:04.278       ...
00:11:04.278  in ./test/compress/compress.sh:91 -> main(["qat"])
00:11:04.278       ...
00:11:04.278     86  	run_bdevperf 32 4096 3
00:11:04.278     87  	run_bdevperf 32 4096 3 512
00:11:04.278     88  	run_bdevperf 32 4096 3 4096
00:11:04.278     89  	
00:11:04.278     90  	if [ $RUN_NIGHTLY -eq 1 ]; then
00:11:04.278  => 91  		run_bdevio
00:11:04.278     92  		run_bdevperf 64 16384 30
00:11:04.278     93  	
00:11:04.278     94  		# run perf on nvmf target w/compressed vols
00:11:04.278     95  		export TEST_TRANSPORT=tcp && nvmftestinit
00:11:04.278     96  		nvmfappstart -m 0x7
00:11:04.278       ...
00:11:04.278  
00:11:04.278  ========== Backtrace end ==========
00:11:04.278    09:11:50	-- common/autotest_common.sh@1032 -- # xtrace_enable
00:11:04.278    09:11:50	-- common/autotest_common.sh@1033 -- # return 0
00:11:04.278   09:11:50	-- compress/compress.sh@1 -- # killprocess 257954
00:11:04.278   09:11:50	-- common/autotest_common.sh@816 -- # '[' -z 257954 ']'
00:11:04.278   09:11:50	-- common/autotest_common.sh@820 -- # kill -0 257954
00:11:04.278    09:11:50	-- common/autotest_common.sh@821 -- # uname
00:11:04.278   09:11:50	-- common/autotest_common.sh@821 -- # '[' Linux = Linux ']'
00:11:04.278    09:11:50	-- common/autotest_common.sh@822 -- # ps --no-headers -o comm= 257954
00:11:04.537   09:11:50	-- common/autotest_common.sh@822 -- # process_name=reactor_0
00:11:04.537   09:11:50	-- common/autotest_common.sh@826 -- # '[' reactor_0 = sudo ']'
00:11:04.537   09:11:50	-- common/autotest_common.sh@834 -- # echo 'killing process with pid 257954'
00:11:04.537  killing process with pid 257954
00:11:04.537   09:11:50	-- common/autotest_common.sh@835 -- # kill 257954
00:11:04.537   09:11:50	-- common/autotest_common.sh@840 -- # wait 257954
00:11:05.913   09:11:51	-- compress/compress.sh@1 -- # error_cleanup
00:11:05.913   09:11:51	-- compress/compress.sh@15 -- # rm -rf /tmp/pmem
00:11:05.913   09:11:51	-- compress/compress.sh@16 -- # /var/jenkins/workspace/crypto-autotest/spdk/build/examples/perf -q 1 -o 131072 -w write -t 2
00:11:05.913  EAL: No available 1048576 kB hugepages reported
00:11:05.913  TELEMETRY: No legacy callbacks, legacy socket not created
00:11:12.475  Initializing NVMe Controllers
00:11:12.475  Attached to NVMe Controller at 0000:87:00.0 [8086:0a54]
00:11:12.475  Associating PCIE (0000:87:00.0) NSID 1 with lcore 0
00:11:12.475  Initialization complete. Launching workers.
00:11:12.475  ========================================================
00:11:12.475                                                                             Latency(us)
00:11:12.475  Device Information                     :       IOPS      MiB/s    Average        min        max
00:11:12.475  PCIE (0000:87:00.0) NSID 1 from core  0:    8471.50    1058.94     117.73      57.05    1871.96
00:11:12.475  ========================================================
00:11:12.475  Total                                  :    8471.50    1058.94     117.73      57.05    1871.96
00:11:12.475  
00:11:12.475   09:11:57	-- compress/compress.sh@1 -- # exit 1
00:11:12.475    09:11:57	-- common/autotest_common.sh@964 -- # trap - ERR
00:11:12.475    09:11:57	-- common/autotest_common.sh@964 -- # print_backtrace
00:11:12.475    09:11:57	-- common/autotest_common.sh@992 -- # [[ ehxBET =~ e ]]
00:11:12.475    09:11:57	-- common/autotest_common.sh@994 -- # args=('qat' './test/compress/compress.sh' 'compress_qat' 'autorun-spdk.conf')
00:11:12.475    09:11:57	-- common/autotest_common.sh@994 -- # local args
00:11:12.475    09:11:57	-- common/autotest_common.sh@996 -- # xtrace_disable
00:11:12.475    09:11:57	-- common/autotest_common.sh@14 -- # '[' no '!=' yes ']'
00:11:12.476    09:11:57	-- common/autotest_common.sh@15 -- # PREV_BASH_OPTS=ehxBET
00:11:12.476    09:11:57	-- common/autotest_common.sh@16 -- # [[ ehxBET == *\x* ]]
00:11:12.476    09:11:57	-- common/autotest_common.sh@17 -- # XTRACE_DISABLED=yes
00:11:12.476    09:11:57	-- common/autotest_common.sh@19 -- # set +x
00:11:12.476  ========== Backtrace start: ==========
00:11:12.476  
00:11:12.476  in /var/jenkins/workspace/crypto-autotest/spdk/test/common/autotest_common.sh:964 -> run_test(["compress_qat"],["./test/compress/compress.sh"],["qat"])
00:11:12.476       ...
00:11:12.476     959 		timing_enter $test_name
00:11:12.476     960 		echo "************************************"
00:11:12.476     961 		echo "START TEST $test_name"
00:11:12.476     962 		echo "************************************"
00:11:12.476     963 		xtrace_restore
00:11:12.476  => 964 		time "$@"
00:11:12.476     965 		xtrace_disable
00:11:12.476     966 		echo "************************************"
00:11:12.476     967 		echo "END TEST $test_name"
00:11:12.476     968 		echo "************************************"
00:11:12.476     969 		timing_exit $test_name
00:11:12.476       ...
00:11:12.476  in /var/jenkins/workspace/crypto-autotest/spdk/autotest.sh:312 -> main(["autorun-spdk.conf"])
00:11:12.476       ...
00:11:12.476     307 		if [ $SPDK_TEST_VMD -eq 1 ]; then
00:11:12.476     308 			run_test "vmd" ./test/vmd/vmd.sh
00:11:12.476     309 		fi
00:11:12.476     310 	
00:11:12.476     311 		if [ $SPDK_TEST_REDUCE -eq 1 ]; then
00:11:12.476  => 312 			run_test "compress_qat" ./test/compress/compress.sh "qat"
00:11:12.476     313 			run_test "compress_isal" ./test/compress/compress.sh "isal"
00:11:12.476     314 		fi
00:11:12.476     315 	
00:11:12.476     316 		if [ $SPDK_TEST_OPAL -eq 1 ]; then
00:11:12.476     317 			run_test "nvme_opal" ./test/nvme/nvme_opal.sh
00:11:12.476       ...
00:11:12.476  
00:11:12.476  ========== Backtrace end ==========
```

**Comment by changpe1** [comment]
[Bug Scrub] let's see if this issue can happen again or not for a while.

**Comment by kgodzwox** [comment]
This issue happens every nightly-build:
https://ci.spdk.io/results/autotest-nightly/builds/696/archive/crypto-autotest/build.log
https://ci.spdk.io/results/autotest-nightly/builds/695/archive/crypto-autotest/build.log
https://ci.spdk.io/results/autotest-nightly/builds/694/archive/crypto-autotest/build.log

etc.

**Comment by karlatec** [comment]
> see if this issue can happen again or not for a while.

@changpe1 from what I can tell this has 100% reproduction rate and hits in every nightly test. This is not an intermittent failure.

**Comment by jimharris** [comment]
Could you do a test run on this with master + a revert of a6e022463?

**Comment by jimharris** [comment]
> Could you do a test run on this with master + a revert of [a6e0224](https://github.com/spdk/spdk/commit/a6e02246348377082d5e7de14d73fab239be8380)?

Please also try master (without any reverts) and this patch:  https://review.spdk.io/gerrit/c/spdk/spdk/+/12672

**Comment by kgodzwox** [comment]
I can confirm that the patch fixes issue.
https://ci.spdk.io/public_build/autotest-nightly_699.html
(crypto passed)

**Comment by karlatec** [comment]
@kgodzwox @tomzawadzki  should we cherry-pick this fix to v22.01.x?
